orangecontrib/prototypes/widgets/utils/histogram.py

```diff
@@ -0,0 +1,305 @@
+            self.n_bins = len(self.attribute.values)
+        elif self.attribute.is_continuous:
+            self.n_bins = n_bins
+
+        # Handle target variable index
+        self.class_index = class_index
+        if self.class_index is not None:
+            assert 0 <= class_index < len(data.domain.class_vars), \
+                'Target variable `%d` out of range!' % class_index
+            self.target_var = data.domain.class_vars[class_index]
+            self.y = data.Y[:, class_index] if data.Y.ndim > 1 else data.Y
+        else:
+            self.target_var, self.y = None, None
+
+        # _plot_`dim` accounts for all the paddings and spacings
+        self._plot_height = self.height - top_padding
+        self._plot_width = self.width - 2 * side_padding - (self.n_bins - 2) * bar_spacing
+
+        self.__layout = QGraphicsLinearLayout(Qt.Horizontal, self)
+        self.__layout.setContentsMargins(side_padding, top_padding, side_padding, 0)
+        self.__layout.setSpacing(bar_spacing)
+
+        if border is not None:
+            self.border = QGraphicsRectItem(0, 0, self.width, self.height, self)
+            self.border.setPen(QPen(QColor(border)))
+
+        self.edges, self.distributions = self._histogram()
+
+        self._draw_histogram()
+
+    def _get_histogram_edges(self):
+        """Get the edges in the histogram based on the attribute type.
+        
+        In case of a continuous attribute, we split the variable range into
+        n bins. In case of a discrete attriubute, bins don't make sense, so we
+        just return the attribute values.
+        
+        This will return the staring and ending edge, not just the edges in
+        between.
+        
+        """
+        if self.attribute.is_discrete:
+            return np.array(self.attribute.values)
+        else:
+            edges = np.linspace(self.x.min(), self.x.max(), self.n_bins)
+            edge_diff = edges[1] - edges[0]
+            return np.hstack((edges, [edges[-1] + edge_diff]))
+
+    def _get_bin_distributions(self, bin_indices):
+        """Compute the distribution of instances within bins.
+
+        Parameters
+        ----------
+        bin_indices : np.ndarray
+            An array with same shape as `x` but containing the bin index of the
+            instance.
+
+        Returns
+        -------
+        np.ndarray
+            A 2d array; first dim represents different bins, second dim the
+            counts of different target values-
+
+        """
+        if self.target_var and self.target_var.is_discrete:
+            y = one_hot(self.y)
+            bins = np.arange(self.n_bins)[:, np.newaxis]
+            mask = bin_indices == bins
+            distributions = np.zeros((self.n_bins, y.shape[1]))
+            for bin_idx in range(self.n_bins):
+                distributions[bin_idx] = y[mask[bin_idx]].sum(axis=0)
+        else:
+            distributions = np.bincount(bin_indices.astype(np.int64))
```

Comment:
Please use our own implementation of `bincount` (`Orange/statistics/util.py`) which can also handle sparse data sets.